lib/publiccloud/basetest.pm

```diff
@@ -152,6 +153,13 @@ sub _cleanup {
 
 sub post_fail_hook {
     my ($self) = @_;
+    if (get_var('PUBLIC_CLOUD_SLES4SAP')) {
```

Comment:
I would suggest to go one level deeper and add this into `_cleanup` . for example you will be able to stop passing cleanup_called and simplify a bit your function

Reply:
That's what I initially tried. Problem is, `_cleanup` is called in the `post_run_hook` as well as the `post_fail_hook` for some reason, and it led to destroyed infrastructure after the module (successfully) finished:

```
sub post_run_hook {
    my ($self) = @_;
    $self->_cleanup() unless $self->{cleanup_called};
}
```

So, I opted to put it directly inside the `post_fail_hook`, since we only want this to run in case of module failure.

Reply:
> That's what I initially tried. Problem is, _cleanup is called in the post_run_hook as well as the post_fail_hook for some reason, and it led to destroyed infrastructure after the module (successfully) finished


this has good reasons behind .  `_cleanup()` method needs to do cleanup not only when test fail but also when test fail , but ofc. we need support the fact that normally tests may have more than one module and you don't want to re-create infra in every module from scratch . We have answer for this too https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/basetest.pm#L130 . You may add `publiccloud_multi_module` flag to test flags of your module and than `_cleanup` will skip actual cleanup call during successful .

Ofc. your current approach will work but I would say that this is step into direction "keep SAP as separate world " which also has some advantages but main and IMHO huge disadvantage of which is **no ability to combine our efforts and need to solve same problems two times** ...

Reply:
Ah, ok, now I understand the logic.For example, you don't set the flag in `ssh_interactive_end`, which is the reason it also executes cleanup. But you do use it in `ssh_interactive_start` or `instance_overview`, which are sure to have more modules after them.

Wouldn't it be better if the flag was reversed? I mean, if there are no other modules, then set a flag to execute cleanup `post_run`, otherwise only execute cleanup on failure, as is the generally expected behavior.

No matter - I will place our cleanup in `_cleanup`, after the flag check, and re-run VRs.

Reply:
> Wouldn't it be better if the flag was reversed? I mean, if there are no other modules, then set a flag to execute cleanup post_run, otherwise only execute cleanup on failure, as is the generally expected behavior.

this was done in current way due to "historical reasons" ( initially we started Public Cloud tests as one-module tests so it was not a problem and later we just add support for "multi-module tests" ) . As amount of single-module tests is "more than one " :) I don't see big difference in one default or another . 
But also I don't if someone will do huge work of changing this default in dozen of places and will run dozens of tests to prove that it is working :)

Reply:
Update: unfortunately, this wont do, as `_cleanup` sets `$self->{cleanup_called} = 1;` at the begining of the function, and as we're using the variable too inside `sles4sap_cleanup`, it leads to the cleanup not being executed.

i can of course explicitly set `cleanup_called` to 0 again before calling our cleanup, but it feels somewhat "hacky". I'd leave it in the `post_fail_hook`, and look to homogenize our code parts as a dedicated task, once we decide to do so (for example when the PC squad decides to use the peering cleanup).

What do you think?